### PR TITLE
Add command line option to limit dataset years

### PR DESCRIPTION
**Pull request by anaschwendler**
This is just a template to help you make your point clear with this PR. :)

**What is the purpose of this Pull Request?**
Following @irio #60 with the option to limit dataset years, this PR:
- Update the way to select an specific path to save the datasets
- Fetch the `Dataset()` method for year, only if `years` is no None
- Update the README.md to this new option

**What was done to achieve this purpose?**
Understand what @irio has done to achieve that new option and fix what was missing to make this new option works

**How to test if it really works?**
Run rosie with a few options like:
```console
$ python rosie.py run chamber_of_deputies --path /another/folder --years 2017
$ python rosie.py run chamber_of_deputies --path /another/folder
$ python rosie.py run chamber_of_deputies --years 2017,2016
$ python rosie.py run chamber_of_deputies 

$ python rosie.py run federal_senate --path /another/folder --years 2017
$ python rosie.py run federal_senate --path /another/folder
$ python rosie.py run federal_senate --years 2017,2016
$ python rosie.py run federal_senate 
```

**Who can help reviewing it?**
@cuducos

### CODE REVIEW

**Comment by anaschwendler** [comment]
There will be some tests to add to rosie, but I think that it is a great start!

**Comment by cuducos** [comment]
> There will be some tests to add to rosie, but I think that it is a great start!

Sure it is! Many thanks for that ; )

Two doubts:

1. Should we wait for these tests before reviewing the PR?
2. As the CLI gets more complicates I think we should adopt either [`argparse`](https://docs.python.org/3/library/argparse.html) (⭐️ ), [`click`](http://click.pocoo.org) (⭐️ ⭐️⭐️⭐️) or [`docopt`](https://github.com/docopt/docopt) ( ⭐️⭐️⭐️⭐️⭐️). What do you think of implementing that in this PR?

**Comment by anaschwendler** [comment]
> Should we wait for these tests before reviewing the PR?

This feature is important, so I must consider accept as it is now, and open a new PR to that

> As the CLI gets more complicates I think we should adopt either argparse (⭐️ ), click (⭐️ ⭐️⭐️⭐️) or docopt ( ⭐️⭐️⭐️⭐️⭐️). What do you think of implementing that in this PR?

I would like to do it now, but I think it is too much for one PR, but I can add too, or we can open an issue and I can do it following, what do you think @cuducos ?

**Comment by cuducos** [comment]
> This feature is important, so I must consider accept as it is now, and open a new PR to that

If the feature is importante I'd say the opposite: let's _not_ accept non-tested code ; ) Also, if the feature is important I think using `argparse`, `click` or `docopt` is very much welcomed: they are highly tested packages to avoid us reinventing the wheel and accidentally getting something wrong in in the way.

**TLDR** I honestly prefer tests and a high level CLI development tool before reviewing this PR bizarrely for the same reason you prefer to postpone tests and high level CI dev tool hahaha…

**Comment by anaschwendler** [comment]
Oh! Nice point, working on it!! \o\

**Comment by anaschwendler** [comment]
So considering what we need to do:
- [ ] Test `chamber_of_deputies` with and without years
- [ ] Test `federal_senate` with and without years
- [ ] Include `docopt` to our CLI

**Comment by anaschwendler** [comment]
Now that I'm full recovered, I will get back to it!

**Comment by anaschwendler** [comment]
Just included `docopt` to our CLI, and had some doubts about documentation.

Docopt works fine, but when I run `python rosie.py` only, it gets a `help >` terminal, is it normal?

image

Independent of what I do, it don't change this option, what can be done about it?

**Comment by anaschwendler** [comment]
Just add the tests with and without years for `chamber_of_deputies`. The result is:

```console
(serenata_rosie) ➜  rosie git:(anaschwendler-limit-years) ✗ python rosie.py test
..............................................................
----------------------------------------------------------------------
Ran 62 tests in 152.857s

OK
```

I'm completely sure that this it not the best way to do that, but I think that the logic is ok. I'll wait for @cuducos review, to see if he can help me with my tests.

File: rosie.py [hunk 1]
```diff
@@ -1,4 +1,20 @@
...
+
+Usage:
+    python rosie.py run (chamber_of_deputies | federal_senate)
+    python rosie.py run (chamber_of_deputies | federal_senate) [--years 2017,2016]
+    python rosie.py run (chamber_of_deputies | federal_senate) [--path <path to output directory>]
+    python rosie.py run (chamber_of_deputies | federal_senate) [--path <path to output directory>] [--years 2017,2016]
+    python rosie.py test [chamber_of_deputies | federal_senate]
+
+Options:
+    --years   runs rosie wuth years filter
```
Comment (cuducos) [inline rosie.py]: Typo: `wuth` -> `with`

Rosie is a proper noun and should be capitalized.

In _years filter_ the word _years_ is an adjective (I guess) and should not be pluralized.

Also _runs rosie wuth years filter_ doesn't tell much about what _years filter_ does. What about: _runs Rosie fetching data and finding suspicions for a specific set of years_?

File: README.md [hunk 2]
```diff
@@ -52,7 +52,12 @@ A `/tmp/serenata-data/irregularities.xz` file will be created. It's a compacted
 Also a target directory (where files are saved) can de passed — for example:
 
 ```console
-$ python rosie.py run chamber_of_deputies /my/serenata/directory/
+$ python rosie.py run chamber_of_deputies --path /my/serenata/directory/
+```
+
+Besides that, if you want to, you can run the project with years filter - for example:
```
Comment (cuducos) [inline README.md]: Line 52 uses a [m-dash](http://www.thepunctuationguide.com/em-dash.html) (instead of an hyphen) before _for example_. This line uses an hyphen. Either we go with m-dash or with hyphen, but swinging between both is weird.

Also a line break between text and mono-spaced block makes the raw markdown way easier to read.

File: rosie.py [hunk 3]
```diff
@@ -1,4 +1,20 @@
...
+    --path    runs rosie with specific directory
+
+"""
+
 from sys import argv, exit
+from docopt import docopt
```
Comment (cuducos) [inline rosie.py]: PEP8 suggests a blank line between native modules (`sys`) and third-party modules (`docopt`).

File: rosie.py [hunk 4]
```diff
@@ -1,4 +1,20 @@
...
+Usage:
+    python rosie.py run (chamber_of_deputies | federal_senate)
+    python rosie.py run (chamber_of_deputies | federal_senate) [--years 2017,2016]
+    python rosie.py run (chamber_of_deputies | federal_senate) [--path <path to output directory>]
+    python rosie.py run (chamber_of_deputies | federal_senate) [--path <path to output directory>] [--years 2017,2016]
+    python rosie.py test [chamber_of_deputies | federal_senate]
+
+Options:
+    --years   runs rosie wuth years filter
+    --path    runs rosie with specific directory
```
Comment (cuducos) [inline rosie.py]: Rosie is a proper noun and should be capitalized.

Also runs _runs rosie with specific directory_ doesn't tell much about what this directory is for What about: _directory in which Rosie saves fetched data and the suspicions CSV_?

If I'm not wrong this optional argument has a default (`data/`? `/tmp/serenata-data/`?) that should be documented as well.

File: rosie.py [hunk 5]
```diff
@@ -26,12 +31,19 @@ def run():
...
         target_module = argv[2]
     else:
-        print('A module must be provided.')
-        help()
+        arguments = docopt(__doc__, help=True)
+        print(arguments)
         exit(1)
-    target_directory = argv[3] if len(argv) >= 4 else '/tmp/serenata-data/'
+
+    target_directory = argv[argv.index('--path') + 1] if '--path' in argv else '/tmp/serenata-data/'
```
Comment (cuducos) [inline rosie.py]: We should use `docopt` for the default value too. From their docs:

> If you want to set a default value for an option with an argument, put it into the option's description, in the form `[default: <the-default-value>]`.

File: rosie.py [hunk 6]
```diff
@@ -26,12 +31,19 @@ def run():
     if len(argv) >= 3:
```
Comment (cuducos) [inline rosie.py]: Any special reason are we are still dealing with `argv` when using `docopt`?

File: rosie.py [hunk 7]
```diff
@@ -26,12 +31,19 @@ def run():
...
         target_module = argv[2]
     else:
-        print('A module must be provided.')
-        help()
+        arguments = docopt(__doc__, help=True)
```
Comment (cuducos) [inline rosie.py]: Why parsing arguments with `docopt` here and not in main? Isn't `docopt` used for both `run` and `test`?

What about parsing arguments else where (`main`?) and calling `run` or `test` when arguments are already parsed?

File: rosie/chamber_of_deputies/tests/test_adapter.py [hunk 8]
```diff
@@ -27,6 +30,18 @@ def setUp(self):
     def tearDown(self):
         shutil.rmtree(self.temp_path)
 
+    def test_quantity_of_downloaded_datasets_all_years_fetch(self):
+        years = AVAILABLE_YEARS
+        chamber_of_deputies = Dataset(self.temp_path, years)
+        datasets = chamber_of_deputies.fetch()
+        self.assertEqual(10, len(datasets))
```
Comment (cuducos) [inline rosie/chamber_of_deputies/tests/test_adapter.py]: This hard-coded `10` will break tests every new year. `len(AVAILABLE_YEARS)` wont ; )

File: rosie/chamber_of_deputies/tests/test_adapter.py [hunk 9]
```diff
@@ -27,6 +30,18 @@ def setUp(self):
...
+
+    def test_filter_years_fetch(self):
+        years = [2017,2016]
```
Comment (cuducos) [inline rosie/chamber_of_deputies/tests/test_adapter.py]: PEP8 recommends a space after this comma.

File: rosie/chamber_of_deputies/tests/test_adapter.py [hunk 10]
```diff
@@ -27,6 +30,18 @@ def setUp(self):
...
+        chamber_of_deputies = Dataset(self.temp_path, years)
+        datasets = chamber_of_deputies.fetch()
+        self.assertEqual(3, len(datasets))
```
Comment (cuducos) [inline rosie/chamber_of_deputies/tests/test_adapter.py]: Why we expect `3` when `len(years) == 2`? Can we document that? For example:

```python
self.assertEqual(3, len(datasets))  # data from 2 years + suspicions file
```

File: rosie.py [hunk 11]
```diff
@@ -1,4 +1,20 @@
...
+    python rosie.py run (chamber_of_deputies | federal_senate) [--path <path to output directory>] [--years 2017,2016]
+    python rosie.py test [chamber_of_deputies | federal_senate]
+
+Options:
+    --years   runs rosie wuth years filter
+    --path    runs rosie with specific directory
+
+"""
+
 from sys import argv, exit
```
Comment (cuducos) [inline rosie.py]: The idea of `docopt` is to avoid using `argv` (`exit` is still needed for the test command).

File: rosie.py [hunk 12]
```diff
@@ -1,4 +1,20 @@
...
+from docopt import docopt
 
 
 def entered_command(argv):
```
Comment (cuducos) [inline rosie.py]: We don't need this method when using `docopt`.

File: rosie.py (hunk 1 above)
Comment (anaschwendler) [reply rosie.py]: accepted 👍

File: rosie.py (hunk 4 above)
Comment (anaschwendler) [reply rosie.py]: > If I'm not wrong this optional argument has a default (data/? /tmp/serenata-data/?) that should be documented as well.

Yes, `serenata-data`, can you help me documenting that? There is something in the `README.md` but I guess we can do more.

Comment (cuducos) [reply rosie.py]: >  can you help me documenting that?

From [the snippet of `docopt` docs I pasted yesterday](https://github.com/datasciencebr/rosie/pull/115#discussion_r159600820) it's is as simple as replacing this very line we're commenting on. Just change:

```
    --path    runs rosie with specific directory
```

by:


```
    --path    runs rosie with specific directory [default: /tmp/serenata-data]
```

Comment (anaschwendler) [reply rosie.py]: accepted

File: rosie/chamber_of_deputies/tests/test_adapter.py (hunk 8 above)
Comment (anaschwendler) [reply rosie/chamber_of_deputies/tests/test_adapter.py]: accepted, I just didn't know we could do that

**Comment by anaschwendler** [comment]
Hey @cuducos I think I got something here :)

I'm sure that there is a few things to fix, but its getting better, can you review for me?

File: rosie.py [hunk 13]
```diff
@@ -1,49 +1,64 @@
...
-    if len(argv) >= 3:
-        target_module = argv[2]
-    else:
-        print('A module must be provided.')
-        help()
-        exit(1)
-    target_directory = argv[3] if len(argv) >= 4 else '/tmp/serenata-data/'
+    target_directory = args['<path>'] if args['--path'] else '/tmp/serenata-data/'
+
+    target_module = 'chamber_of_deputies' if args['chamber_of_deputies'] else 'federal_senate'
```
Comment (anaschwendler) [inline rosie.py]: @cuducos I'm not into this way of selecting `target_module` do you have an better option for that?

Comment (cuducos) [reply rosie.py]: When using `docopt` we should have in `arguments` a dictionary like this:

```python
{'--path': False,
 '--years': False,
 'chamber_of_deputies': True,
 'core': False,
 'federal_senate': False,
 'run': True,
 'test': False}
```

So I'd go for something like:

```python
commands = ('chamber_of_deputies', 'federal_senate')
target_module, *_ =  filter(lambda x: arguments.get(x), commands)
```

File: rosie.py [hunk 14]
```diff
@@ -1,49 +1,64 @@
...
+
     klass = getattr(rosie, target_module)
-    klass.main(target_directory)
 
+    if args['--years']:
+        years = args['<years>']
+        years = [int(num) for num in years.split(',')]
+        klass.main(target_directory, years=years)
+    else:
+        klass.main(target_directory)
```
Comment (cuducos) [inline rosie.py]: No need for this `if`/`else` here since:

* if `args['--years']` is `None`
* and considering that the default value for `years` is `None` (and it is)
* then `klass(target_dir)` is exactly equal to `klass(target_dir, args['--years']`

However I'm afraid we might need to convert the string `2016,2017` to a list of integers, i.e. `[2016, 2017]`. For example:

```python
if args['--years']:
   args['--years'] = [int(v) for v in args['--years'].split(',')]

klass.main(target_directory, args['--years'])
```

File: rosie.py [hunk 15]
```diff
@@ -1,49 +1,64 @@
...
-        target_module = argv[2]
-        tests_path = os.path.join('rosie', target_module)
+    if args['chamber_of_deputies']:
+        tests_path = os.path.join('rosie', 'chamber_of_deputies')
+        tests = loader.discover(tests_path)
+    elif args['federal_senate']:
+        tests_path = os.path.join('rosie', 'federal_senate')
         tests = loader.discover(tests_path)
     else:
         tests = loader.discover('rosie')
```
Comment (cuducos) [inline rosie.py]: Minor DRY suggestion. What about:

```python
test_path = 'rosie'
if args['chamber_of_deputies']:
     tests_path = os.path.join(test_path, 'chamber_of_deputies')
elif args['federal_senate']:
     tests_path = os.path.join(test_path, 'federal_senate')
tests = loader.discover(test_path)
```

Comment (anaschwendler) [reply rosie.py]: I still think that this is not good enough :/
maybe another way to test this with `@patch`?

File: rosie.py [hunk 16]
```diff
@@ -1,49 +1,64 @@
...
     import rosie.federal_senate
 
-    if len(argv) >= 3:
-        target_module = argv[2]
-    else:
-        print('A module must be provided.')
-        help()
-        exit(1)
-    target_directory = argv[3] if len(argv) >= 4 else '/tmp/serenata-data/'
+    target_directory = args['<path>'] if args['--path'] else '/tmp/serenata-data/'
```
[the remark this answers is not in the document]
Comment (anaschwendler) [reply rosie.py]: I knew that should be more improvements, thanks for that, looks pretty good :)

**Comment by anaschwendler** [comment]
Hi @cuducos took a closer look on what you have said, and made some adjustments. I'm still not happy with the (test script)[https://github.com/datasciencebr/rosie/pull/115/files#diff-c9a174d8dd8e2ad3e2749ec8267d5aab]. But now the main script I think that may look nice. Can you review again for me?

File: rosie/chamber_of_deputies/tests/test_adapter.py [hunk 17]
```diff
@@ -27,6 +30,18 @@ def setUp(self):
     def tearDown(self):
         shutil.rmtree(self.temp_path)
 
+    def test_quantity_of_downloaded_datasets_all_years_fetch(self):
+        years = AVAILABLE_YEARS
```
Comment (cuducos) [inline rosie/chamber_of_deputies/tests/test_adapter.py]: Why do we need this? You use `years` in line 35 but go back to `AVAILABLE_YEARS` in 37.

**Comment by cuducos** [comment]
>  I'm still not happy with the test script.

What bugs you in this script?

File: rosie/chamber_of_deputies/tests/test_adapter.py (hunk 17 above)
Comment (anaschwendler) [reply rosie/chamber_of_deputies/tests/test_adapter.py]: didn't noticed, thanks /o/

**Comment by anaschwendler** [comment]
@cuducos It takes time by downloading the datasets from the source, I can't think of anything, but maybe there is another way to test that, without downloading the whole datasets

**Comment by cuducos** [comment]
> there is another way to test that, without downloading the whole datasets

You could mock HTTP requests, or you could mock only the classes (as internally the way `self.years` works might be already tested).
